core/misc/slot_chain.go

```diff
@@ -0,0 +1,180 @@
+package misc
+
+import (
+	"sync"
+
+	"github.com/alibaba/sentinel-golang/core/base"
+	"github.com/alibaba/sentinel-golang/core/log"
+	"github.com/alibaba/sentinel-golang/core/stat"
+)
+
+var (
+	rsSlotChainLock sync.RWMutex
+	rsSlotChain     = make(map[string]*base.SlotChain, 8)
+
+	globalStatPrepareSlots = make([]base.StatPrepareSlot, 0, 8)
+	globalRuleCheckSlots   = make([]base.RuleCheckSlot, 0, 8)
+	globalStatSlot         = make([]base.StatSlot, 0, 8)
+)
+
+func GlobalStatPrepareSlots() []base.StatPrepareSlot {
+	return globalStatPrepareSlots
+}
+
+func GlobalRuleCheckSlots() []base.RuleCheckSlot {
+	return globalRuleCheckSlots
+}
+
+func GlobalStatSlot() []base.StatSlot {
+	return globalStatSlot
+}
+
+// RegisterGlobalStatPrepareSlot is not thread safe, and user must call RegisterGlobalStatPrepareSlot when initializing sentinel running environment
+func RegisterGlobalStatPrepareSlot(slot base.StatPrepareSlot) {
+	for _, s := range globalStatPrepareSlots {
+		if s.Name() == slot.Name() {
+			return
+		}
+	}
+	globalStatPrepareSlots = append(globalStatPrepareSlots, slot)
+}
+
+// RegisterGlobalRuleCheckSlot is not thread safe, and user must call RegisterGlobalRuleCheckSlot when initializing sentinel running environment
+func RegisterGlobalRuleCheckSlot(slot base.RuleCheckSlot) {
+	for _, s := range globalRuleCheckSlots {
+		if s.Name() == slot.Name() {
+			return
+		}
+	}
+	globalRuleCheckSlots = append(globalRuleCheckSlots, slot)
+}
+
+// RegisterGlobalStatSlot is not thread safe, and user must call RegisterGlobalStatSlot when initializing sentinel running environment
+func RegisterGlobalStatSlot(slot base.StatSlot) {
+	for _, s := range globalStatSlot {
+		if s.Name() == slot.Name() {
+			return
+		}
+	}
+	globalStatSlot = append(globalStatSlot, slot)
+}
+
+func newResourceSlotChain() *base.SlotChain {
+	sc := base.NewSlotChain()
+	sc.AddStatPrepareSlotLast(stat.DefaultResourceNodePrepareSlot)
+
+	for _, s := range globalStatPrepareSlots {
+		sc.AddStatPrepareSlotLast(s)
+	}
+
+	for _, s := range globalRuleCheckSlots {
+		sc.AddRuleCheckSlotLast(s)
+	}
+
+	sc.AddStatSlotLast(stat.DefaultSlot)
+	sc.AddStatSlotLast(log.DefaultSlot)
+	for _, s := range globalStatSlot {
+		sc.AddStatSlotLast(s)
+	}
+
+	return sc
+}
+
+func validateStatPrepareSlot(sc *base.SlotChain, s base.StatPrepareSlot) bool {
+	flag := false
+	f := func(slot base.StatPrepareSlot) {
+		if slot.Name() == s.Name() {
+			flag = true
+		}
+	}
+	sc.RangeStatPrepareSlot(f)
+
+	return flag
+}
+
+func RegisterResourceStatPrepareSlot(rsName string, slot base.StatPrepareSlot) {
+	rsSlotChainLock.Lock()
+	defer rsSlotChainLock.Unlock()
+
+	sc, ok := rsSlotChain[rsName]
+	if !ok {
+		sc = newResourceSlotChain()
+	}
+
+	if !validateStatPrepareSlot(sc, slot) {
+		sc.AddStatPrepareSlotLast(slot)
+	}
+	if !ok {
+		rsSlotChain[rsName] = sc
```

Comment:
when ok equals true , register new slot, rsSlotChain[rsName] will not be updated.

Reply:
sc is a pointer(*SlotChain), update is unnecessary.

Reply:
It might be better to do so.
```
func RegisterResourceRuleCheckSlot(rsName string, slot base.RuleCheckSlot) {
	rsSlotChainLock.Lock()
	defer rsSlotChainLock.Unlock()

	sc, ok := rsSlotChain[rsName]
	if !ok {
		sc = newResourceSlotChain()
                rsSlotChain[rsName] = sc
	}

	if !validateRuleCheckSlot(sc, slot) {
		sc.AddRuleCheckSlotLast(slot)
	}
}
```